### ISSUE TITLE
Where you do locate the LIWC English dictionary once you purchase the LIWC license?

### ISSUE THREAD

**Issue by illushka** [chbrown/liwc-python#5]
I see that you need to purchase an LIWC license to get access to the dictionary. So, I have purchased the software but I can’t figure out how to access the English Dictionary file which your Python package needs for it to work. Any pointers would be really helpful as to how to connect your python package to the LIWC dictionary easily e.g. where do I access/download the LIWC English file?

**Comment by d0tN3t** [comment]
Hey, were you able to get it figured out? @illushka

**Comment by illushka** [comment]
No. Far as I gather after speaking to LIWC directly, this version of LIWC is now outdated.

**Comment by d0tN3t** [comment]
Pfft...outdated? It took them from 2007 to 2015 to make their "current" one.
Email me at jason.quintana1@gmail.com I'll show you how.

**Comment by maifeng** [comment]
For LIWC 2015, the dictionary in the following .jar file in LIWC installation folder (e.g. C:\Program Files (x86)\LIWC2015\): `\app\lib\LIWC2015-app-1.4.0.jar`.

You can open the jar file using WinRAR as an archive and find `LIWC2015_English_Flat.dic` in  `com/liwc/LIWC2015/data/dict/`. 

As mentioned in #3 , you need to edit the  __init__.py file to read the dictionary. Namely, you need to replace `parts = tsv.split()` with `parts = tsv.split("\t")`. An alternative solution is to do a global search and remove all spaces in the LIWC2015_English_Flat.dic file. **Caveat**: using either solution will not consider phrases such as "to like" and "kind of" in the LIWC 2015 dictionary.

**Comment by d0tN3t** [comment]
![Screen Shot 2019-02-15 at 12 51 17 PM](https://user-images.githubusercontent.com/11362484/54088807-e3730280-432f-11e9-8588-56fc47bc7835.png)

lol...I was going to say the same thing, but I don't know the legalities of that so I just assumed not to say anything, but since you already said it :)

**Comment by meettyj** [comment]
![image](https://user-images.githubusercontent.com/30004852/71633103-3fb95280-2be0-11ea-8728-c392e4375d9d.png)

You can still get access to the dict files, but the content is meaningless. File name: LIWC2015-app-1.6.0

**Comment by d0tN3t** [comment]
That is incorrect. Here is a screenshot of said dic-file converted.  It's several thousand rows in length.
![image](https://user-images.githubusercontent.com/11362484/72759571-b0c6a700-3b9b-11ea-8813-d365752f12f4.png)

**Comment by ma-ji** [comment]
> That is incorrect. Here is a screenshot of said dic-file converted. It's several thousand rows in length.
> ![image](https://user-images.githubusercontent.com/11362484/72759571-b0c6a700-3b9b-11ea-8813-d365752f12f4.png)

I don't know how you get this, but the file on my computer is encrypted.

**Comment by d0tN3t** [comment]
I can't tell if you're trolling me or not lol 🤣. I did it MacOS originally, but I just tried @maifeng method and it worked too. If you're not finding it my only guess is maybe because I bought the full value one rather than the academic...but I distinctly remember doing the little trick on MacOS just the academic version. It wasn't till way later that I bought the full.
Maybe Vscode has a plugin for reading dic files...or maybe try opening it in Excel or Google Sheets. 

dic flat_file

![image](https://user-images.githubusercontent.com/11362484/74855171-f2b34c00-5305-11ea-93bb-6de05172898e.png)

regular dic file
![image](https://user-images.githubusercontent.com/11362484/74855425-4d4ca800-5306-11ea-8cdb-5a7fc65aa9d5.png)

**Comment by maifeng** [comment]
In 1.6.0 they encoded the dict file. You can download an older version (or decode it if you study the source code).

**Comment by ma-ji** [comment]
Thanks @d0tN3t , I hard-coded it anyway.

**Comment by d0tN3t** [comment]
@ma-ji 
It says on your profile that you're at UT-Austin. All that being said you could ask the man responsible for LIWC, the legend, James W Pennebaker. He teaches at your school. Ha-Ha, but I doubt he would be pleased we're craking the software dictionary. 🤣

**Comment by ma-ji** [comment]
@d0tN3t They have a protocol, so in principle, we should follow for sure :)

**Comment by shayneqian** [comment]
> In 1.6.0 they encoded the dict file. You can download an older version (or decode it if you study the source code).

I purchased from liwc.net recently and also got the encrypted dictionary files (extracted from LIWC2015-app-1.6.0.jar). Would you mind sharing some more information on how to decode the .dic files? It seems that it is impossible to download an older version from the links provided in the purchasing mail.

**Comment by d0tN3t** [comment]
@shayneqian Hmmm... so I just checked the link where I'm able to redownload my purchase and it still shows the options (my only options actually) for 1.5.0. 
If you look at the bottom left corner of my image it shows the version after hovering over Windows 64-bit. 
Maybe different purchases are tied to different links?
![Screen Shot 2020-03-05 at 6 17 25 AM](https://user-images.githubusercontent.com/11362484/75981318-e43e6600-5ea9-11ea-9cec-8b9a7df6cd1d.png)

**Comment by newtomtl83** [comment]
I have the same issue. The dictionary is encrypted or something. Any suggestion?

**Comment by d0tN3t** [comment]
@newtomtl83 

Hey sorry to hear that. Hmm....can you get to the poster? It looks something like below and it may even have the name "poster" in it. You'll have to do some digging but it should be in PDF format.
![image](https://user-images.githubusercontent.com/11362484/79080828-dddbaf00-7cdd-11ea-9a88-86b87c9e1fe5.png)

**Comment by newtomtl83** [comment]
> @newtomtl83
> 
> Hey sorry to hear that. Hmm....can you get to the poster? It looks something like below and it may even have the name "poster" in it. You'll have to do some digging but it should be in PDF format.
> ![image](https://user-images.githubusercontent.com/11362484/79080828-dddbaf00-7cdd-11ea-9a88-86b87c9e1fe5.png)

This is great, I found it! Any suggestion on how to transform it into a text doc? I tried copy pasting the content into excel but it doesn't work.

**Comment by d0tN3t** [comment]
Wow, you're quick. Yes! You'll have to do some organizing if you're up for the challenge but run this code on the pdf after pip installing tika.

```
from tika import parser

raw = parser.from_file('LIWC2015.pdf')
print(raw['content'])
```

You should get something like this.
![image](https://user-images.githubusercontent.com/11362484/79081376-6a886c00-7ce2-11ea-8c41-c324f7fbec0a.png)

**Comment by d0tN3t** [comment]
@newtomtl83 Finally, when you get to making your .dic file there is a format that needs to be followed and I was only able to get the chbrown liwc library (liwc 0.3.0) to recognize both my custom LIWC file and the production value one. 

Below is my custom LIWC file. The format (if I remember correctly) is:
```
%"\n"
1"\t"prepositions"\n"
2"\t"quantifiers"\n"
. . .
%"\n"
abandon"\t"5"\n"
abandoned"\t"5"\n"
```
So basically replace empty spaces with tab seperation(tsv) and end each line with "new-line" delimitation.

![image](https://user-images.githubusercontent.com/11362484/79081898-17fd7e80-7ce7-11ea-9caf-c219165588ba.png)

**Comment by sara-02** [comment]
@d0tN3t Sorry for this question but I too have LIWC 2015 and I tried the above to steps to extract the dict form pdf, but the pdf extractor does not give us the mapping for word belonging to which column. How to obtain that?
And what does each newline represent? One column?

**Comment by d0tN3t** [comment]
@sara-02 Hey!

Yes, unfortunately, it's not an easy task so you will have to be creative but there is definitely a pattern that you can find to build up the dictionary.

So when Tika does the extraction it contains the format. The format is going across from the 1st column **("1 Function")** to the last **("125 Filler")** and the words are contained in what looks like a paragraph followed by "new line" which ends the row. So you're reading the group of tokens by rows. **The newline equals the end of the row.**
![pstr_1](https://user-images.githubusercontent.com/11362484/79902039-69d49180-83d6-11ea-9a86-65b95217202e.PNG)

The mapping (numbers/category) are towards the bottom. Those are what you will be using to number each word in the dictionary. 
![pstr_2](https://user-images.githubusercontent.com/11362484/79902192-ac966980-83d6-11ea-96a4-058f018536fb.PNG)

As for the mapping, below is the example of the dictionary. You can't see it but "\t" and "\n" are delimiting the dictionary's format.  So let's look at the 2nd and 3rd word `"abondon*" & "abort*"`. These words (like many others) are overlapping and fall under multi-categories. So when building the dictionary, you'll want to reference the poster for which column the words fall under.  This is what it looks like:
```
abandon*"\t"12"\t"16"\t"19"\t"20"\t"24"\n"
abort*"\t"60"\t"61"\t"62"\n"
```
![pstr_3](https://user-images.githubusercontent.com/11362484/79902710-9210c000-83d7-11ea-9839-29d59df10375.PNG)

I honestly don't think it makes a difference but for what it is worth, the numbers assigned to the category names are below. Like I've said in my previous post, I've made my own dictionary and numbered the way I wanted. 
![image](https://user-images.githubusercontent.com/11362484/79902918-ddc36980-83d7-11ea-8155-e1713ff90970.png)

**Comment by d0tN3t** [comment]
@sara-02 Ok, soooo maybe this might work. Looking at one of my previous posts, I tried the link in the snapshot and it automatically downloaded the 1.5.0 version.
![image](https://user-images.githubusercontent.com/11362484/79904587-8672c880-83da-11ea-9c69-d963e4552166.png)
This is the link:
`http://liwc.net/liwcdownload/LIWC2015apps/LIWC2015-x64-perm-1.5.0.exe`

Let me know if this works...it's a long shot, but I just tried it and it automatically downloaded it. And if you can download this version then you just have to follow the steps above for extracting the .dic file.  Best of luck! 🙏👍

**Comment by fris-t** [comment]
@d0tN3t 
Thanks for the suggestion. Just tried it, and they have the same encryption...
Sad!

**Comment by sara-02** [comment]
> @d0tN3t
> Thanks for the suggestion. Just tried it, and they have the same encryption...
> Sad!

Yes, same for me.

**Comment by d0tN3t** [comment]
@fris-t  @sara-02  Ugh...alright then...sorry everyone. 😌

**Comment by virresh** [comment]
After struggling with this for quite a while, I finally understood the encryption scheme.
Here is a POC binary that should help y'all decrypt the encrypted files. It's built for Linux x64 distros. In case you don't have one, you can probably try it out inside docker / on google colab / or a cloud VM.

Do note that it doesn't have any error checks, so if you give in garbage, you'll either get error, or garbage output. Extract the binary from .zip file before using it (had to zip it because direct binary attachments are not supported).
Usage: ./binary <encrypted_input.dic> <decrypted_output.dic>

[decrypter.zip](https://github.com/chbrown/liwc-python/files/4533757/decrypter.zip)

**Comment by chbrown** [comment]
@virresh presuming good faith, I appreciate the contribution; I haven't run your program or otherwise dealt with this encryption issue, but might I ask you provide the source code? Running an anonymous binary is dangerous, and providing the binary without the source is more than a tad ironic in this particular situation.

**Comment by virresh** [comment]
@chbrown thanks for the added disclaimer. I understand the concern, however, I cannot make the source public without exposing the key used for the above encryption. I feel comfortable sharing binary because it has similar obfuscation as the original LIWC (I'm not sure about legalities, but at least not I'm not exposing anything more than already has been).

Since I cannot share the exact source, I'm sharing a redacted version, whoever likes can use this after extracting the key from LIWC package:
https://gist.github.com/virresh/e965753c39fa641d20aaeeab96f04c76

**Comment by chbrown** [comment]
Ah, interesting. Thanks for the gist. I'm no C++ expert, but looks like they just XOR the original dic with a cycled key, byte by byte? That's... cute :)

I'm working in the dark here since I don't have a copy of one of these new-fangled encrypted LIWC versions; can you provide any instructions for other users on where to find the key in their LIWC package?

If it's somewhere predictable, it'd be easy enough to do the decryption on the fly in Python implementation, where instead of providing just the path to the `.dic` file, you'd also provide the path to the key (using your C++ implementation as reference).

**Comment by fris-t** [comment]
@virresh Thank you so much for trying to crack the code. 
I am relatively advanced in my Python skills, but in Linux I'am still a noob.
As such, I really like @chbrown's suggestion. I am willing to help crafting a part of the code if necessary! 
Hope to hear from you!

**Comment by virresh** [comment]
> Ah, interesting. Thanks for the gist. I'm no C++ expert, but looks like they just XOR the original dic with a cycled key, byte by byte? That's... cute :)

Yep, that's pretty much all they're doing.

> I'm working in the dark here since I don't have a copy of one of these new-fangled encrypted LIWC versions; can you provide any instructions for other users on where to find the key in their LIWC package?
> 
> If it's somewhere predictable, it'd be easy enough to do the decryption on the fly in Python implementation, where instead of providing just the path to the `.dic` file, you'd also provide the path to the key (using your C++ implementation as reference).

@fris-t @chbrown, I figure code will speak about the process more than anything. So, here's a reference script for python:
https://gist.github.com/virresh/dab626aed9dd7933c3d3ee0300042f19

Note that in case the library is updated again, one would have to decompile the files from scratch and study the code to crack the new encryption schemes in place.

**Comment by fris-t** [comment]
@virresh Thanks so much. I am going to give this a shot right away...I'll get back to you

**Comment by bensilver95** [comment]
> @ma-ji
> I can't tell if you're trolling me or not lol 🤣. I did it on MacOS originally, but I just tried @maifeng method. And it worked too. If you're not finding it my only guess is maybe because I bought the full value one rather than the academic...but I distinctly remember doing the little trick on MacOS just the academic version. It wasn't till way later that I bought the full.
> Maybe Vscode has a plugin for reading dic files...or maybe try opening it in Excel or Google Sheets.
> 
> dic flat_file
> 
> ![image](https://user-images.githubusercontent.com/11362484/74855171-f2b34c00-5305-11ea-93bb-6de05172898e.png)
> 
> regular dic file
> ![image](https://user-images.githubusercontent.com/11362484/74855425-4d4ca800-5306-11ea-8cdb-5a7fc65aa9d5.png)

I have LIWC2015 academic version on a mac and I am unable to even locate the original .dic file. I see on this thread it's mentioned how to find it on Windows, but I don't see anything about where it is located on a mac. Can anyone point me to where it is?

**Comment by bensilver95** [comment]
@d0tN3t I have the LIWC2015 academic version on a mac and I am unable to even locate the original .dic file. I see that you posted how to find it on Windows, and you mentioned that you had located the .dic file with the academic version on a mac. Can anyone point me to where it is?

**Comment by fris-t** [comment]
Go to your applications folder. Normally you have the LIWC2015 program installed there.
Then right click on the pictogram and choose 'show contents of package' or something like that. I have a French OS, so don't know the english term (see picture below)
![Capture d’écran 2020-04-30 à 20 25 11](https://user-images.githubusercontent.com/45532009/80745510-b3cd1e00-8b20-11ea-9443-dc9088b617a0.png)

And then it is in this .jar file, over here.
![Capture d’écran 2020-04-30 à 20 26 20](https://user-images.githubusercontent.com/45532009/80745621-dcedae80-8b20-11ea-9473-7c35c1f2333c.png)

I used the java decompiler program (http://java-decompiler.github.io/) to decompose the jar.

**Comment by bensilver95** [comment]
@fris-t Thanks, that worked swimmingly. But is there a way to copy the .dic file to another folder? I don't need to read it, I just need to copy it. Within JD-GUI, it does not seem that I'm able to copy it to somewhere else, or call on its location

**Comment by d0tN3t** [comment]
@bensilver95 Hey!

Yes @fris-t  is correct. That is how you get to the jar file. Cntrl click (right-click) on the Liwc2015 icon > show package content > Jave > lib > LIWC2015.app-1.6.0.jar. Then you need to make a copy of it, drop it in a new folder (desktop), and use WinRar to extract it. 

@maifeng  pretty much summed it up by writing:
"You can open the jar file using WinRAR as an archive and find LIWC2015_English_Flat.dic in com/liwc/LIWC2015/data/dict/."

**Comment by d0tN3t** [comment]
@fris-t  Hey!

Just a heads up, I'm not sure if those parentheses will cause any errors as the encryption version doesn't have them next to the word.

![image](https://user-images.githubusercontent.com/11362484/80749330-e4df2b80-8aeb-11ea-8a0a-5000abb868b2.png)

![image](https://user-images.githubusercontent.com/11362484/80749579-40a9b480-8aec-11ea-9cfa-0d2c3ae0fb02.png)


Also, does yours have more categories than the original one?
Mine has a total of 68, I believe

**Comment by fris-t** [comment]
@d0tN3t hmmm, Interesting observation....Maybe it is because it is a screenshot of the not-flat .dic? 

I have also decrypted the flat dictionary and that looks like this:

![Capture d’écran 2020-04-30 à 21 46 20](https://user-images.githubusercontent.com/45532009/80752598-09f38e80-8b2c-11ea-89e6-86553a679795.png)
![Capture d’écran 2020-04-30 à 21 46 43](https://user-images.githubusercontent.com/45532009/80752623-17a91400-8b2c-11ea-9cc4-04e87e54c81e.png)
![Capture d’écran 2020-04-30 à 21 47 21](https://user-images.githubusercontent.com/45532009/80752675-2e4f6b00-8b2c-11ea-9622-750fc6c55d9a.png)

There seem to be 125 categories in this dictionary. Is that possible? Could it be like that this is an updated version or so (this .dic came with the latest version of LIWC (1.6.0))? Or do you have the impression there are errors/doubles?

**Comment by virresh** [comment]
My 5 cents:
The categories are only 73

Observe the s.no on the editor
The category numbers are probably chosen arbitrarily (or maybe some complicated bit logic)

**Comment by d0tN3t** [comment]
@fris-t 
Nope, nevermind, I forgot about the non-flat dic.
![image](https://user-images.githubusercontent.com/11362484/80773848-c6475780-8b20-11ea-8aed-fb1cc81e848f.png)

**Comment by xionglong58** [comment]
> ![image](https://user-images.githubusercontent.com/30004852/71633103-3fb95280-2be0-11ea-8728-c392e4375d9d.png)
> 
> You can still get access to the dict files, but the content is meaningless. File name: LIWC2015-app-1.6.0.jar

do you fix it?

**Comment by chbrown** [comment]
Hi all, after being politely contacted by the LIWC developers, I've decided that this line of inquiry is outside the scope of this repository.

I originally started this project because I didn't completely understand how the standard LIWC application was producing its analysis — e.g., how it handles tokenization, since the official application processes entire text documents as input — and doing it myself in Python gave me more control over those types of details, and more transparency as to the precise nature of the output. (It also allowed me to stay in my familiar research environment — a mixture of Python modules and IPython notebooks — rather than open up a GUI and click around whenever I got to that step in the analysis.)

I appreciate the interest, and I'm sure the LIWC developers wholly support all you researchers using LIWC to answer your research questions... legitimately and within the terms of the LIWC license. Because I also appreciate the amount of time and energy that the LIWC developers have put into producing the lexicon, I think it's best to comply with their request, so that they can continue to improve LIWC, and so that us LIWC users can continue to benefit from those improvements. They're good people 🤗 , and I don't want to discredit or disrespect their efforts.

tl;dr: https://github.com/chbrown/liwc-python/#obtaining-liwc

**Comment by chbrown** [comment]
![tldr-you-dont-meme](https://github.com/user-attachments/assets/b73a017b-7e24-4082-aeb7-8213c72eb249)

LIWC developers reached out again and asked me to delete details / comments here, which I have now done.

They point out that this effort is "in contravention of Section 8 of the LIWC Terms of Use":

> “you are not allowed to modify, reverse engineer, decompile, disassemble or use for any derivative works the LIWC program or any of the LIWC components for any purpose.”
